src/sort.ts

```diff
@@ -0,0 +1,94 @@
+/** @public */
+export type SortDirection =
+  | 1
+  | -1
+  | 'asc'
+  | 'desc'
+  | 'ascending'
+  | 'descending'
+  | { $meta: string };
+
+/** @public */
+export type Sort =
+  | string
+  | string[]
+  | { [key: string]: SortDirection }
+  | [string, SortDirection][]
+  | [string, SortDirection];
+
+/** Below stricter types were created for sort that correspond with type that the cmd takes  */
+
+/** @internal */
+type SortDirectionForCmd = 1 | -1 | { $meta: string };
+
+/** @internal */
+type SortForCmd = { [key: string]: SortDirectionForCmd };
+
+/** @internal */
+class SortDigest {
+  static prepareDirection(direction: any = 1): SortDirectionForCmd {
+    const value = ('' + direction).toLowerCase();
+    if (SortDigest.isMeta(direction)) return direction;
+    switch (value) {
+      case 'ascending':
+      case 'asc':
+      case '1':
+        return 1;
+      case 'descending':
+      case 'desc':
+      case '-1':
+        return -1;
+      default:
+        throw new Error(`Invalid sort direction: ${JSON.stringify(direction)}`);
+    }
+  }
+  static isMeta(t: SortDirection): t is { $meta: string } {
+    return typeof t === 'object' && t !== null && '$meta' in t && typeof t.$meta === 'string';
+  }
+  static isPair(t: Sort): t is [string, SortDirection] {
+    if (Array.isArray(t) && t.length === 2) {
+      try {
+        SortDigest.prepareDirection(t[1]);
+        return true;
+      } catch (e) {
+        return false;
+      }
+    }
+    return false;
+  }
+  static pairToObject(v: [string, SortDirection]): SortForCmd {
+    return { [v[0]]: SortDigest.prepareDirection(v[1]) };
+  }
+  static isDeep(t: Sort): t is [string, SortDirection][] {
+    return Array.isArray(t) && Array.isArray(t[0]);
+  }
+  static deepToObject(t: [string, SortDirection][]): SortForCmd {
+    return t.reduce((acq, i) => {
+      return { ...acq, ...SortDigest.pairToObject(i) };
+    }, {});
```

Comment:
It's unlikely that anyone will have large sort specifiers but nonetheless this nested looping feels like something to avoid despite the potentially small input.

Reply:
I'm confused, this seems essential, how else would you convert `[['a', 1], ['b', 'desc']]` to `{a: 1, b: -1}`?

Reply:
Wouldn't this be equivalent without the inner loop?
```ts
let sortObject = {};
for (const [name, value] of t) {
  sortObject[name] = prepareDirection(value);
}
return sortObject
```

Reply:
cool, done